### ISSUE TITLE
Non-deterministic workflow failure from slot deployment

### ISSUE THREAD

**Issue by mitchellrust** [Azure/azure-functions-durable-extension#2696]
### Description
**Microsoft Support Case Tracking ID:** `2310310040011442`

This was previously opened as bug #2635 but it was auto-closed.

We have seen occasional failures in our eternal durable orchestrations during deployment to our function apps, using slot deployments. This does not occur on every deployment, but we've seen it repeated multiple times.

The error message is as follows:

`Non-Deterministic workflow detected: A previous execution of this orchestration scheduled an activity task with sequence ID 0 and name 'DetectAndPublishSalesOrderEvents' (version ''), but the current replay execution hasn't (yet?) scheduled this task. Was a change made to the orchestrator code after this instance had already started running?`

Here is the scenario that we see this happening in:
- Our function app uses slot deployments to minimize downtime during deployments, with two slots: `stage` and `production`.
- During said deployments where this issue occurs, no changes to orchestrations or activities have been made.
- Each of our slots have their own task hubs, with their own control queues and tables. These hubs are contained in the same Azure Storage Account. We've defined this in our deployment app settings for the `production` and `stage` slots individually, using the `AzureFunctionsJobHost__extensions__durableTask__hubName` sticky app setting for each.
- During a slot deployment, there could be any number of orchestrations in the middle of executing on our `production` slot, or waiting for scheduled activity tasks to complete before resuming. We utilize eternal orchestrations heavily for our app.
  - We have seen this failure exclusively when an orchestration is awaiting the completion of a scheduled activity task.
- During the "Slot Swap" step of our slot deployment, we have seen in logs that orchestrations awaiting activity completions get resumed in the `stage` slot. This doesn't make a ton of sense, as we wouldn't expect any tasks to be scheduled on the `stage` slot control queues, but perhaps something in the app-settings switch causes things to cross over here.

Based on our scenario and what the logs have shown us, it seems that our `stage` slot is resuming orchestrations somehow, and naturally can't find the history for the orchestration because that all exists in the `production` slot task hub. This seems to be nuance specific to using the slot deployments, from what we can tell.

### Expected behavior
A slot swap during deployment does not resume orchestrations and their related request-messages within the `stage` slot. 

### Actual behavior
After the slot swap has occurred during deployment, the `stage` slot resumes eternal orchestrations that have been running in the `production` slot, previous to the slot deployment. As expected, the `stage` slot cannot find the orchestration history, and fails with `Non-Deterministic workflow detected` error.

### Relevant source code snippets
N/A

### Known workarounds
None. Microsoft support has communicated that we should stop all running orchestrations before executing a slot deployment, however this is counter to our use of eternal orchestrations and is not deemed a workaround.

### App Details

- **Durable Functions extension version (e.g. v1.8.3)**: Microsoft.Azure.WebJobs.Extensions.DurableTask v2.9.0
- **Azure Functions runtime version (1.0 or 2.0)**: Runtime Version v4.27.5.21549
- **Programming language used**: C# Dotnet 6.0

### Screenshots
`production` slot app configuration, showing task hub sticky deployment setting.
![image](https://github.com/Azure/azure-functions-durable-extension/assets/23386919/e321c771-242d-467b-8762-03a6d3044174)

`stage` slot app configuration, showing task hub sticky deployment setting.
![image](https://github.com/Azure/azure-functions-durable-extension/assets/23386919/a82bdccb-ae53-4184-9f58-720ea3eeadce)

### If deployed to Azure
- **Timeframe issue observed**: `2023-12-07T21:06:15.4034258Z` to  `2023-12-07T21:08:21.3668316Z`
- **Function App name**: `gsp01fap-jde-na-ext-prod`
- **Function name(s)**: `MonitorSalesOrderChanges` (orchestration), `DetectAndPublishSalesOrderEvents`  (activity)
- **Azure region**: `West US`
- **Orchestration instance ID(s)**: `monitor$MonitorSalesOrderChanges$2a6408375de940aaa756f613c6ecd42e`
- **Azure storage account name**: `gsp01strjdenaext`
```[tasklist]
### Tasks
```

**Comment by bachuv** [comment]
Hi @mitchellrust, have you had a chance to review our [Zero downtime deployment](https://learn.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-zero-downtime-deployment) docs? It covers scenarios when using slots is recommended. If you have long running orchestrations, you can try the application routing strategy. Let us know if you have any questions!

**Comment by drdamour** [comment]
@bachuv yes we've read through that quite a bit.  Does the application routing strategy apply to eternal orchestration singletons?  it seems to apply to incoming requests to launch orchestrations...but that's not the case with eternal orchestrations, they just keep running.  It also seems to really focus on breaking changes, but in this case we are NOT deploying breaking changes, yet the eternal orchestration crashes.

**Comment by bcrispcvna** [comment]
We're running into the same issue while moving a function app from .NET 6 in-proc to .NET 8 isolated.  We also have a stage/production slot and all of our in-flight eternal orchestrations will fail during the swap with a similar exception.  

This is currently blocking us from moving forward with the upgrade.  I'm going to try a workaround for now to disable our orchestration function(s) on the prod slot prior to the swap and re-enabling it afterwards.  It's not a preferred scenario so hopefully this is something that can be addressed.

@mitchellrust @bachuv is this something that you've tried already?

**Comment by mitchellrust** [comment]
@bcrispcvna this isn't something that we've tried, we've just been restarting any of the orchestrations that randomly fail when performing the slot swap. We are hoping to not build too much process around a workaround in hopes that a proper fix can be put in place for the bug.

**Comment by nytian** [comment]
Hey @mitchellrust  If I understand correctly, production slot and stage slot has its own Task Hub which won't be swapped. I wonder, is there any orchestrator instances that is in Pending or Running state at stage slot's TaskHub when the swap happened? If there is, that would cause the instances resume running. And if the orchestrator logic changes, that would trigger the non-deterministic exception.

**Comment by drdamour** [comment]
@nytian theres no change in orchestrator logic

**Comment by nytian** [comment]
Do you have any recent app that experience this issue( less than 6 month) so that I can check logs?

Also, for zero-downtime deployment, we recently introducing `Versioning` feature please check here https://learn.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-orchestration-versioning?tabs=csharp

**Comment by drdamour** [comment]
i think this can be closed @mitchellrust .  but oh my, what a confusing situation! that i'm sure others have hit but weren't able to take the time (it took us near a year with MS support) to find the cause of.  If you came upon this issue...i hope this explanation is clear enough.

we found the root problem to be that there are windows of time during a slot swap where managed identity is switched, but config is not.  this was confirmed through MS support case.  So the sequence is like:
1. swap started
2. stage slot gets prod slots managed identity config and restarts
3. stage slot gets prod config and restarts
4. prod slot gets stage's managed identity config and restarts
5. prod slot gets stage's config and restarts
6. swap complete

we had coded our activities to detect and fail in the time between 2&3 and 4&5 so nothing bad would happen during a config/identity "mismatch".   THIS IS THE PART MOST FOLKS PROBABLY DON'T REALIZE that they can run with a "mismatched" identity and config for a brief window.  It's intermittent because your orchestration must start up and run or an activity must be worked during these usually pretty brief (less than a minute) windows.

NOTE: there was SOME mention by the MS rep that they may flip 4 & 5's order...i don't know if that happened.  also this isn't special for function apps, it's same for all app services so your MVC app may be using the wrong identity for brief windows...though it won't be receiving http traffic....it's the queue triggered stuff like orchestrations and activities that will bite you.

Our orchestration had been coded to retry activities that failed so we kept thinking we were good, but kept getting non-deterministic workflow.  Finally it was realized the failure wasn't in the orchestration method...but in the constructor of the class of the orchestration method!!!  The non-determinism applies to the constructor too!  an invoke of the orchestration method if it's an instance method is create the class instance then call the orchestration method. well..that constructor was taking DI service dependencies (for the activities) and one of them deep down failed if the identity and config were mistmatched (specifically a call to keyvault can't resolve if you have an identity for a different keyvault). And yeah...in hindsight duh all the service resolution is actually part of the constructor and thus ALSO falls into the non-determinism constraint.


our new best practice is to ALWAYS have orchestration methods be static and NEVER instance methods even if the class is filled with instance activity methods that the orchestration invokes.  this makes it much clearer what's happening and we avoid all DI possibilities in orchestration methods.

Maybe the azf docs should have some stronger wording about why orchestration methods should always be preferred to be static and not instance methods so others might miss this DI trap...our DI was only non-deterministic because of the swap behaviour...but we could have avoided it all together if we knew just a bit more about how hot the fire we were holding really was.  Heck i'd even suggest going so far as making in a roslyn analzyer warning that you must supress if you don't want it.
